### DISCUSSION TITLE
What are Vault and its helpers?

### DISCUSSION THREAD

**Discussion by bencomp** [IIIF-Commons/vault#21] [Q&A]
After clicking a link in Slack to a vault helper, then to its repository root, then finding this Vault repository, I couldn't find a description of what I'm looking at. What are Vault and the Vault helpers?

**Comment by stephenwf** [comment]
There is a good overview here: https://iiif-canvas-panel.netlify.app/docs/components/vault and there is also some un-styled, very early documentation for Vault here: https://iiif-vault.netlify.app/docs (written up to Reading resources)

Vault is a normalisation layer and store for IIIF. Any version of IIIF goes in, and a predictable subset of IIIF 3.0 comes out. Additionally, it flattens IIIFs nested structures. 

This allows for passing around of "references" or `{id, type}` pairs around an application, instead of large-full objects. It also allows for tools and helpers to be built on top of slices of IIIF, instead of full manifests. It also makes subscribing to and making changes in a deeply nested IIIF document easier. 

Vault helpers are a collection of small utilities. If they depend on deeply nested IIIF - they do this through vault as a dependency. For example, there is a Painting annotations helper:

```ts
import { Vault } from '@iiif/vault'; 
import { createPaintingAnnotationsHelper } from '@iiif/vault-helpers';

const vault = new Vault();
const helper = createPaintingAnnotationsHelper(vault);

await vault.load('https://example.org/manifest-1');

// get all painting annotations for a canvas
const annotations = helper.getAllPaintingAnnotations('https://example.org/canvas-1');

// Extract choices (if any) from a canvas
const choices = helper.extractChoices('https://example.org/canvas-1');
```
 
Vault tries to stay as close to IIIF Presentation 3.0 as possible and avoids introducing opinion in how data should be abstracted. Vault helpers add opinionated helpers and custom data-structures to make working with some aspects of IIIF easer.
